userspace/async/async_key_value_source.h

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright (C) 2019 Sysdig, Inc.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#pragma once
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <list>
+#include <map>
+#include <mutex>
+#include <set>
+#include <thread>
+#include <stdint.h>
+
+namespace sysdig
+{
+
+/**
+ * Base class for classes that need to collect values asynchronously from some
+ * value source.  Subclasses will override the the run_impl() method and
+ * implement the concrete value lookup behavior.  In that method, subclasses
+ * will use use dequeue_next_key() method to get the key that it will use to
+ * collect the value(s), collect the appropriate value(s), and call the
+ * store_value() method to save the value.  The run_impl() method should
+ * continue to dequeue and process values while the dequeue_next_key() method
+ * returns true.
+ *
+ * The constructor for this class accepts a maximum wait time; this specifies
+ * how long client code is willing to wait for a synchronous response (i.e.,
+ * how long the lookup() method will block waiting for the requested value).
+ * If the async_key_value_source is able to collect the requested value
+ * within that time period, then the lookup() method will return them.
+ *
+ * If the lookup() method is unable to collect the requested value within
+ * the requested time period, then one of two things will happen.
+ *
+ * <ol>
```

Comment:
will browsers actually render this? even if not, it's still amusing and an effective way to do it. ascii art lovers everywhere bang their heads against a wall

Reply:
Yeah, Doxygen allows for embedded HTML, and it'll include it in the generated documentation.